FinalClass should exempt private-only constructor classes that also have nested subclasses

```bash
$ cat Test.java
```

```java
public class Test {

    public static final Test ONE = new Test() {
        @Override
        public int value() {
            return 1;
        }
    };

    private Test() {
    }

    public int value() {
        return 0;
    }
}
```

```bash
$ cat config.xml
```

```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="FinalClass"/>
    </module>
</module>
```

```bash
$ RUN_LOCALE="-Duser.language=en -Duser.country=US"
$ java $RUN_LOCALE -jar checkstyle-8.41-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /Users/archie/bugs/checkstyle-A/Test.java:1:1: Class Test should be declared as final. [FinalClass]
Audit done.
Checkstyle ends with 1 errors.
```

The `FinalClass` module is doing it's job correctly in that it's detecting a non-final class with only private constructor(s), but it's not smart enough to realize that in some situations - e.g., this particular situation where you are defining nested classes within the outer class - it's not possible to make the class `final` because otherwise those nested classes would be impossible to write, because the compile would fail because you can't extend a final class.

Java allows nested classes to extend a class using a private superclass constructor. So the upshot is that `FinalClass` should make an exception when the class being checked contains any nested subclasses.

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java
@@ -80,6 +80,15 @@ import com.puppycrawl.tools.checkstyle.utils.ScopeUtil;
  *     private MyClass() { }
  *   }
  * }
+ *
+ * class TestAnonymousInnerClasses { // OK, class has an anonymous inner class.
+ *     public static final TestAnonymousInnerClasses ONE = new TestAnonymousInnerClasses() {
+ *
+ *     };
+ *
+ *     private TestAnonymousInnerClasses() {
+ *     }
+ * }
  * </pre>
  * <p>
  * Parent is {@code com.puppycrawl.tools.checkstyle.TreeWalker}
@@ -128,7 +137,12 @@ public class FinalClassCheck
 
     @Override
     public int[] getRequiredTokens() {
-        return new int[] {TokenTypes.CLASS_DEF, TokenTypes.CTOR_DEF, TokenTypes.PACKAGE_DEF};
+        return new int[] {
+            TokenTypes.CLASS_DEF,
+            TokenTypes.CTOR_DEF,
+            TokenTypes.PACKAGE_DEF,
+            TokenTypes.LITERAL_NEW,
+        };
     }
 
     @Override
@@ -168,6 +182,17 @@ public class FinalClassCheck
                 }
                 break;
 
+            case TokenTypes.LITERAL_NEW:
+                if (ast.getFirstChild() != null
+                        && ast.getLastChild().getType() == TokenTypes.OBJBLOCK) {
+                    for (ClassDesc classDesc : classes) {
+                        if (doesNameOfClassMatchAnonymousInnerClassName(ast, classDesc)) {
+                            classDesc.registerAnonymousInnerClass();
+                        }
+                    }
+                }
+                break;
+
             default:
                 throw new IllegalStateException(ast.toString());
         }
@@ -178,7 +203,8 @@ public class FinalClassCheck
         if (ast.getType() == TokenTypes.CLASS_DEF) {
             final ClassDesc desc = classes.pop();
             if (desc.isWithPrivateCtor()
-                && !desc.isDeclaredAsAbstract()
+                && !(desc.isDeclaredAsAbstract()
+                    || desc.isWithAnonymousInnerClass())
                 && !desc.isDeclaredAsFinal()
                 && !desc.isWithNonPrivateCtor()
                 && !desc.isWithNestedSubclass()
@@ -218,6 +244,20 @@ public class FinalClassCheck
                 }
             }
         }
+    }
+
+    /**
+     * Check if class name matches with anonymous inner class name.
+     *
+     * @param ast current ast.
+     * @param classDesc class to match.
+     * @return true if current class name matches anonymous inner
+     *         class name.
+     */
+    private static boolean doesNameOfClassMatchAnonymousInnerClassName(DetailAST ast,
+                                                               ClassDesc classDesc) {
+        final String[] className = classDesc.getQualifiedName().split("\\.");
+        return ast.getFirstChild().getText().equals(className[className.length - 1]);
     }
 
     /**
@@ -326,6 +366,9 @@ public class FinalClassCheck
         /** Does class have nested subclass. */
         private boolean withNestedSubclass;
 
+        /** Does class have anonymous inner class. */
+        private boolean withAnonymousInnerClass;
+
         /**
          *  Create a new ClassDesc instance.
          *
@@ -364,6 +407,11 @@ public class FinalClassCheck
         /** Adds nested subclass. */
         private void registerNestedSubclass() {
             withNestedSubclass = true;
+        }
+
+        /** Adds anonymous inner class. */
+        private void registerAnonymousInnerClass() {
+            withAnonymousInnerClass = true;
         }
 
         /**
@@ -409,6 +457,15 @@ public class FinalClassCheck
          */
         private boolean isDeclaredAsAbstract() {
             return declaredAsAbstract;
+        }
+
+        /**
+         * Does class have an anonymous inner class.
+         *
+         * @return true if class has anonymous inner class
+         */
+        private boolean isWithAnonymousInnerClass() {
+            return withAnonymousInnerClass;
         }
 
     }

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheckTest.java
@@ -46,7 +46,12 @@ public class FinalClassCheckTest
     @Test
     public void testGetRequiredTokens() {
         final FinalClassCheck checkObj = new FinalClassCheck();
-        final int[] expected = {TokenTypes.CLASS_DEF, TokenTypes.CTOR_DEF, TokenTypes.PACKAGE_DEF};
+        final int[] expected =
+            {TokenTypes.CLASS_DEF,
+             TokenTypes.CTOR_DEF,
+             TokenTypes.PACKAGE_DEF,
+             TokenTypes.LITERAL_NEW,
+            };
         assertArrayEquals(expected, checkObj.getRequiredTokens(),
                 "Default required tokens are invalid");
     }
@@ -56,9 +61,11 @@ public class FinalClassCheckTest
         final DefaultConfiguration checkConfig =
             createModuleConfig(FinalClassCheck.class);
         final String[] expected = {
-            "7:1: " + getCheckMessage(MSG_KEY, "InputFinalClass"),
-            "15:4: " + getCheckMessage(MSG_KEY, "test4"),
-            "113:5: " + getCheckMessage(MSG_KEY, "someinnerClass"),
+            "9:1: " + getCheckMessage(MSG_KEY, "InputFinalClass"),
+            "17:4: " + getCheckMessage(MSG_KEY, "test4"),
+            "115:5: " + getCheckMessage(MSG_KEY, "someinnerClass"),
+            "149:1: " + getCheckMessage(MSG_KEY, "TestNewKeyword"),
+            "182:5: " + getCheckMessage(MSG_KEY, "NestedClass"),
         };
         verify(checkConfig, getPath("InputFinalClass.java"), expected);
     }
@@ -120,7 +127,12 @@ public class FinalClassCheckTest
     @Test
     public void testGetAcceptableTokens() {
         final FinalClassCheck obj = new FinalClassCheck();
-        final int[] expected = {TokenTypes.CLASS_DEF, TokenTypes.CTOR_DEF, TokenTypes.PACKAGE_DEF};
+        final int[] expected =
+            {TokenTypes.CLASS_DEF,
+             TokenTypes.CTOR_DEF,
+             TokenTypes.PACKAGE_DEF,
+             TokenTypes.LITERAL_NEW,
+            };
         assertArrayEquals(expected, obj.getAcceptableTokens(),
                 "Default acceptable tokens are invalid");
     }

--- a/src/test/resources/com/puppycrawl/tools/checkstyle/checks/design/finalclass/InputFinalClass.java
+++ b/src/test/resources/com/puppycrawl/tools/checkstyle/checks/design/finalclass/InputFinalClass.java
@@ -4,6 +4,8 @@
 ////////////////////////////////////////////////////////////////////////////////
 package com.puppycrawl.tools.checkstyle.checks.design.finalclass;
 
+import java.util.ArrayList;
+
 public class InputFinalClass
 {
     private InputFinalClass() {}
@@ -127,3 +129,74 @@ interface TestInterface {
         private SomeClass() {}
     }
 }
+
+class TestAnonymousInnerClasses { // ok
+    public static final TestAnonymousInnerClasses ONE = new TestAnonymousInnerClasses() {
+        @Override
+        public int value() {
+            return 1;
+        }
+    };
+
+    private TestAnonymousInnerClasses() {
+    }
+
+    public int value() {
+        return 0;
+    }
+}
+
+class TestNewKeyword { // violation, class should be declared final.
+
+    private TestNewKeyword(String s) {
+        String a = "hello" + s;
+    }
+
+    public int count() {
+        return 1;
+    }
+    public static final TestNewKeyword ONE = new TestNewKeyword("world");
+    public static final test6 TWO = new test6() {
+    };
+
+    public static void main(String[] args) {
+        ArrayList<String> foo = new ArrayList<>();
+        foo.add("world");
+        foo.forEach(TestNewKeyword::new);
+    }
+}
+
+interface TestNewKeywordInsideInterface { // ok
+    ArrayList<String> foo = new ArrayList<>();
+}
+
+// abstract classes cannot be final
+abstract class TestPrivateCtorInAbstractClasses { // ok
+    private TestPrivateCtorInAbstractClasses() {
+    }
+}
+
+class TestAnonymousInnerClassInsideNestedClass { // ok
+    private TestAnonymousInnerClassInsideNestedClass() { }
+
+    static class NestedClass { // violation, class should be declared final
+
+        public final static TestAnonymousInnerClassInsideNestedClass ONE
+                = new TestAnonymousInnerClassInsideNestedClass() {
+        };
+
+        private NestedClass() {}
+    }
+    static class NestedClass2 { // ok
+
+        private NestedClass2() {}
+
+        public static final test6  ONE = new test6() {
+
+            public final NestedClass2 ONE = new NestedClass2() {
+
+            };
+        };
+    }
+}
+

--- a/src/xdocs/config_design.xml
+++ b/src/xdocs/config_design.xml
@@ -371,6 +371,15 @@ public @interface Test {
     private MyClass() { }
   }
 }
+
+class TestAnonymousInnerClasses { // OK, class has an anonymous inner class.
+    public static final TestAnonymousInnerClasses ONE = new TestAnonymousInnerClasses() {
+
+    };
+
+    private TestAnonymousInnerClasses() {
+    }
+}
         </source>
       </subsection>
 

